### PR TITLE
feat(llmobs): submit langchain tool.invoke tool spans

### PR DESCRIPTION
**Pull request by yahya-mouman**
This PR adds instrumentation to submit langchain tool.invoke/ainvoke tool spans to APM and LLM Observability. 

The tool spans sent to APM will contain the following I/O data and tags :
langchain.request.tool.name
langchain.request.tool.description
langchain.request.tool.metadata
langchain.request.tool.tags
langchain.request.input
langchain.request.config
langchain.response.output

The tool spans sent to LLM Observability will contain the following I/O data:
input.value: a text field annotating the tool input
output.value: The output of the tool (can be any type)

Ticket : https://datadoghq.atlassian.net/browse/MLOB-935

## Checklist
- [ ] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [ ] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/internal/langchain/patch.py [hunk 1]
```diff
@@ -965,6 +965,170 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+                    "langchain.request.tool.handle_validation_error",
+                    integration.trunc(str(instance.handle_validation_error)),
+                )
+            if instance.response_format:
+                span.set_tag_str(
+                    "langchain.request.tool.response_format", integration.trunc(str(instance.response_format))
+                )
+            if tool_input:
+                span.set_tag_str("langchain.request.input", integration.trunc(str(tool_input)))
+            if config:
```
Comment (yahya-mouman) [inline ddtrace/contrib/internal/langchain/patch.py:1014]: Is this enough or should I parse the configs one by one and add them

File: ddtrace/contrib/internal/langchain/patch.py [hunk 2]
```diff
@@ -965,6 +965,170 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+            integration.log(
+                span,
+                "info" if span.error == 0 else "error",
+                "sampled %s.%s.%s" % (func.__module__, func.__class__.__name__, func.__self__.__class__.__name__),
+                attrs={
+                    # Comment : Should we have more info in the log ?
+                    "tool_name": instance.__self__.name or "",
+                    "input": tool_input,
+                    "config": config or "",
+                },
```
Comment (yahya-mouman) [inline ddtrace/contrib/internal/langchain/patch.py:1045]: Should we have more info in the log ?

File: ddtrace/contrib/internal/langchain/patch.py [hunk 3]
```diff
@@ -965,6 +965,170 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+            )
+    return tool_output
+
+
+@with_traced_module
+async def traced_base_tool_ainvoke(langchain, pin, func, instance, args, kwargs):
```
Comment (yahya-mouman) [inline ddtrace/contrib/internal/langchain/patch.py:1051]: Is there a better way to trace the async variant besides duplicating the code ?

File: ddtrace/contrib/internal/langchain/patch.py [hunk 4]
```diff
@@ -965,6 +965,170 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
     return documents
 
 
+@with_traced_module
+def traced_base_tool_invoke(langchain, pin, func, instance, args, kwargs):
```
Comment (yahya-mouman) [inline ddtrace/contrib/internal/langchain/patch.py:969]: I couldn't think of any metrics to emit from this function

File: ddtrace/contrib/internal/langchain/patch.py [hunk 5]
```diff
@@ -965,6 +965,166 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+    integration = langchain._datadog_integration
+    tool_input = get_argument_value(args, kwargs, 0, "input")
+    config = get_argument_value(args, kwargs, 1, "config") if len(args) >= 2 else None
```
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:972]: So from looking at langchain's [code](https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/tools/base.py#L390-L395) it looks like `config` is an optional keyword argument. This means there may be a case where `len(args)` is not 2 but it is present in `kwargs`, but this code will always return `None` in that case.
```suggestion
    config = get_argument_value(args, kwargs, 1, "config", optional=True)
```

File: ddtrace/contrib/internal/langchain/patch.py [hunk 6]
```diff
@@ -965,6 +965,166 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+
+    span = integration.trace(
+        pin,
+        "%s.%s.%s" % (func.__module__, func.__class__.__name__, func.__self__.__class__.__name__),
```
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:976]: I think here I'd prefer to just use the tool class name instead of the entire module/path for simplicity/clarity. WDYT?

File: ddtrace/contrib/internal/langchain/patch.py (hunk 4 above)
Comment (Yun-Kim) [reply ddtrace/contrib/internal/langchain/patch.py:969]: Are you talking about datadog metrics? If so we can ignore, that's out of scope for our team and this integration.

File: ddtrace/contrib/internal/langchain/patch.py [hunk 7]
```diff
@@ -965,6 +965,166 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+        interface_type="tool",
+        submit_to_llmobs=True,
+    )
+
+    tool_output = None
+    try:
+        if instance.name:
+            span.set_tag_str("langchain.request.tool.name", integration.trunc(str(instance.name)))
```
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:984]: Truncation is only ever performed on input/outputs, so no need to do it here

File: ddtrace/contrib/internal/langchain/patch.py [hunk 8]
```diff
@@ -965,6 +965,166 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+        if integration.is_pc_sampled_span(span):
```
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:985]: `pc_sampling` refers to prompt/completion sampling, or inputs/outputs in this case. We can safely tag most of the below items without needing to check `is_pc_sampled_span(...)`, only need this check for `input` and the return value.

File: ddtrace/contrib/internal/langchain/patch.py [hunk 9]
```diff
@@ -965,6 +965,166 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+                )
+            if instance.handle_validation_error:
+                span.set_tag_str(
+                    "langchain.request.tool.handle_validation_error",
+                    integration.trunc(str(instance.handle_validation_error)),
+                )
+            if instance.response_format:
+                span.set_tag_str(
+                    "langchain.request.tool.response_format", integration.trunc(str(instance.response_format))
+                )
```
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1011]: I don't think we need to tag these parameters other than metadata, description, and tags. We can always add more tags in future PRs, but in case this is too much excessive information being tagged on the spans it will be more difficult to get rid of this tagging in the future.

File: ddtrace/contrib/internal/langchain/patch.py (hunk 1 above)
Comment (Yun-Kim) [reply ddtrace/contrib/internal/langchain/patch.py:1014]: 1. We don't need to truncate config as that isn't a direct I/O to the tool
2. We can probably just json dump the config as it is a `TypedDict` ([see source](https://python.langchain.com/v0.2/api_reference/_modules/langchain_core/runnables/config.html#RunnableConfig))

File: ddtrace/contrib/internal/langchain/patch.py [hunk 10]
```diff
@@ -965,6 +965,166 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+            if tool_input:
+                span.set_tag_str("langchain.request.input", integration.trunc(str(tool_input)))
+            if config:
+                span.set_tag_str("langchain.request.config", integration.trunc(str(config)))
+        tool_output = func(*args, **kwargs)
+        if tool_output is not None:
+            span.set_tag_str("langchain.response.output", integration.trunc(str(tool_output)))
+    except Exception:
+        span.set_exc_info(*sys.exc_info())
+        integration.metric(span, "incr", "request.error", 1)
```
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1021]: No need for adding datadog metrics here, I think this is out of scope for this work and our team.

File: ddtrace/contrib/internal/langchain/patch.py [hunk 11]
```diff
@@ -965,6 +965,166 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
...
+            integration.log(
+                span,
+                "info" if span.error == 0 else "error",
+                "sampled %s.%s.%s" % (func.__module__, func.__class__.__name__, func.__self__.__class__.__name__),
+                attrs={
+                    "tool_name": instance.__self__.name or "",
+                    "input": tool_input,
+                    "config": config or "",
+                },
+            )
```
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1044]: emitting logs is deprecated and discontinued, so can remove this

File: ddtrace/contrib/internal/langchain/patch.py (hunk 3 above)
Comment (Yun-Kim) [reply ddtrace/contrib/internal/langchain/patch.py:1049]: Unfortunately nope 😢 I think there's a couple ways to depollute the patch module (extract tracing functions to separate files, reuse common code in the sync/async functions) but we can keep this for now.

Comment (Yun-Kim) [reply ddtrace/contrib/internal/langchain/patch.py:1049]: Same comments apply to the async function as above

File: releasenotes/notes/feat-llmobs-submit-langchain-tool-spans-d7274ae79d3808e9.yaml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    APM : The LangChain integration now submits spans from tool.invoke to APM as tool spans.
+    LLM Observability: The LangChain integration now submits tool spans from tool.invoke to LLM Observability as retrieval spans.
```
Comment (Yun-Kim) [inline releasenotes/notes/feat-llmobs-submit-langchain-tool-spans-d7274ae79d3808e9.yaml:5]:
```suggestion
    Langchain : Adds support for tracing tool invocations.
    LLM Observability: The LangChain integration now submits tool spans to LLM Observability.
```

File: tests/contrib/langchain/test_langchain_community.py [hunk 13]
```diff
@@ -1274,3 +1274,54 @@ def test_faiss_vectorstore_retrieval(langchain_community, langchain_openai, requ
...
+@pytest.mark.snapshot
+def test_base_tool_invoke(langchain, langchain_core, request_vcr):
+    """
+    Test that invoking a tool with langchain will
+    result in a 1-span trace with a tool span.
+    """
+    from math import pi
+
+    cassette_name = "langchain_tool_invoke.yaml"
+    with request_vcr.use_cassette(cassette_name):
```
Comment (Yun-Kim) [inline tests/contrib/langchain/test_langchain_community.py:1288]: AFAIK tools can be called without a network connection. Why do we need a cassette for this?

File: tests/contrib/langchain/test_langchain_llmobs.py [hunk 14]
```diff
@@ -147,6 +147,15 @@ def _embed_documents(cls, embedding_model, documents, mock_tracer, cassette_name
         LLMObs.disable()
         return mock_tracer.pop_traces()[0]
 
+    @classmethod
+    def _invoke_tool(cls, tool, tool_input, mock_tracer, cassette_name):
+        LLMObs.enable(ml_app=cls.ml_app, integrations_enabled=False, _tracer=mock_tracer)
+        with get_request_vcr(subdirectory_name=cls.cassette_subdirectory_name).use_cassette(cassette_name):
+            if LANGCHAIN_VERSION > (0, 1):
```
Comment (Yun-Kim) [inline tests/contrib/langchain/test_langchain_llmobs.py:154]: Is there any particular reason we're version gating this here?

File: tests/contrib/langchain/test_langchain_llmobs.py [hunk 15]
```diff
@@ -622,6 +631,41 @@ def test_llmobs_embedding_documents(
...
+            name = "Circumference calculator"
+            description = "use this tool when you need to calculate a circumference using the radius of a circle"
+
+            def _run(self, radius):
+                return float(radius) * 2.0 * pi
+
+            def _arun(self, radius: int):
+                raise NotImplementedError("This tool does not support async")
+
+        cassette_name = "langchain_tool_invoke_39.yaml" if PY39 else "langchain_tool_invoke.yaml"
```
Comment (Yun-Kim) [inline tests/contrib/langchain/test_langchain_llmobs.py:650]: same here, not sure if we need a cassette file for this test.

**Review by Yun-Kim — COMMENTED** [review]
Let's split out the APM integration and LLMObs integration into separate PRs for easier readability.

File: ddtrace/contrib/internal/langchain/patch.py (hunk 4 above)
Comment (yahya-mouman) [reply ddtrace/contrib/internal/langchain/patch.py:969]: Don't we use any metrics from those emitted from the langchain integration ?

File: tests/contrib/langchain/test_langchain_community.py (hunk 13 above)
Comment (yahya-mouman) [reply tests/contrib/langchain/test_langchain_community.py]: Yes gotcha ! Was thinking of adding the tool to an LLM instead opted for a simple test. I did remove it in an unpushed commit. Thanks !
I still left the option to call invoke with a cassette in llmobs side for any future agent test to be made.

File: ddtrace/contrib/internal/langchain/patch.py (hunk 9 above)
Comment (yahya-mouman) [reply ddtrace/contrib/internal/langchain/patch.py]: I went for an extensive tracing for all available data but now that you mention it tags and metadata should be the ones giving more information than the rest.

File: tests/contrib/langchain/test_langchain_llmobs.py (hunk 14 above)
Comment (yahya-mouman) [reply tests/contrib/langchain/test_langchain_llmobs.py]: > Note, you'll only need to trace BaseTool.invoke() for langchain V1 and up.
In the ticket description you mentioned this so I was trying to make sure it's the case. I'm guessing since it's in langchain_community test suite it's always going to be the case ?

File: tests/contrib/langchain/test_langchain_community.py [hunk 16]
```diff
@@ -1274,3 +1274,65 @@ def test_faiss_vectorstore_retrieval(langchain_community, langchain_openai, requ
...
+    ignores=["meta.langchain.request.tool.description"],
+    token="tests.contrib.langchain.test_langchain_community.test_base_tool_invoke",
+)
+def test_base_tool_invoke(langchain_core, request_vcr):
+    """
+    Test that invoking a tool with langchain will
+    result in a 1-span trace with a tool span.
+    """
+    if langchain_core is None:
+        pytest.skip("langchain-core not installed which is required for this test.")
```
[the remark this answers is not in the document]
Comment (yahya-mouman) [reply tests/contrib/langchain/test_langchain_community.py:1289]: I agree but the issue is that this specific dependency isn't needed on all tests and if I understand correctly, fixtures are not available on the level of skipif annotations.
So in order to have it available there i'll need to import it on the whole file instead. Which is not very optimal since it's only used on a couple tests.

Do you know of any workaround for this ?

[the remark this answers is not in the document]
Comment (romainkomorndatadog) [reply tests/contrib/langchain/test_langchain_community.py:1289]: I don't know of a workaround if you want to keep using `langchain_core` as a fixture, no.

My personal taste is that it's more traceable to `import langchain_core` in this module and `skipif` if it's absent, than to go sort out why the `langchain_core` fixture is `None`.

Looks like there are ~10 tests that use the `langchain_core` fixture, so it looks like it's going to get loaded in just about all CI cases anyway?

Anyway, it's a very small nit. :)

[the remark this answers is not in the document]
Comment (yahya-mouman) [reply tests/contrib/langchain/test_langchain_community.py:1289]: Yeah i completely understand your point. 
I believe I'll just keep it as it is right now since I haven't authored the file before this PR and wouldn't want to change the existing structure until I speak to the original authors.
Thanks for giving it a look Romain ! 🙇

**Comment by yahya-mouman** [comment]
Split to these two PRs 
PR 1 : https://github.com/DataDog/dd-trace-py/pull/10476
PR 2 (depends on PR1) : https://github.com/DataDog/dd-trace-py/pull/10477
